### DISCUSSION TITLE
[Bug]: Frigate starts writing to local disk if network storage is intermittently not available, does not recover, bricks Home Assistant

### DISCUSSION THREAD

**Discussion by DReffects** [blakeblackshear/frigate#13951] [Report a Bug]
### Checklist

- [X] I have updated to the latest available Frigate version.
- [X] I have cleared the cache of my browser.
- [X] I have tried a different browser to see if it is related to my browser.
- [X] I have tried reproducing the issue in [incognito mode](https://www.computerworld.com/article/1719851/how-to-go-incognito-in-chrome-firefox-safari-and-edge.html) to rule out problems with any third party extensions or plugins I have installed.

### Describe the problem you are having

Two days ago I did some maintenance to my central storage system, therefor the SMB share for frigate became unavailable for a certain amount of time. 
Today, when looking at System Metrics -> Storage the wrong storage device is being used to calculate disk usage.
In my case the Systems internal SSD with 240GB is being shown (Home Assistant installed on x64 NUC), while frigate media is actually being stored on a network share set up within Home Assistant that has 12TB of total space. 

![image](https://github.com/user-attachments/assets/49a7bd3a-e0b6-46db-b151-75038b629a1d)

Further investigation revealed that frigate created the storage file structure on the local disk (/media/frigate) after the SMB share became unavailable. 

This led to the following issues:

- recordings never went back to the SMB share, even after it was back online
- local home assistant disk space was 100% full after two days
- restarting Home Assistant never resulted in mouting of the SMB share to /media/frigate since frigate created a new local file structur at the same path 
- The storage metrics statistics are messed up (minor issue of course)
- Home Assistant was bricked to to 100% disk utilization

This is a very problematic error, since a split second network outage breaks the entirety of video recordings - and Home assistant in the aftermath. While I use a rather large 240GB disk for the HA installation, most people only have very small local storage available so it bricks the system for them even faster.

Since there is no message or notification whatsoever this breaks the entire HA installation due to the disk running full. Plus you loose your recordings from the point forward your disk gets full if you do not notice immediately.

Also recovery from this state is rather problematic as one needs the ability to ssh into the machine and clean up the storage manually to even make HA run again.

I really like that frigate tries to compensate for an outage in the media storage location but I strongly suggest that this becomes configurably in the frigate.yaml config file. 

The outage of centralized storage is not a question of "if" but only "when". You update the firmware of your switch? Outage (with ubiquiti thats every few weeks). You restart the fileserver after an update? Outage. You change out a GBIC module? Outage. You replace a cable in your rack? Outage. You run your stuff via Wifi? Well, outages are preprogrammed then. 

I suggest the following:
- make two storage paths available via config
- primary and secondary
- have frigate check if primary is available periodically
- if not, write to secondary but with a different set of retention to acommedate available diskspace
- issue a notification / warning that the secondary storage is being used
- recover automatically when primary storage is available again and move the secondary storage contents' to the primary

It is my point of view that besides all fancy detecion stuff the most important part of surveillance is a bullet proof method for recording. The system has to recover from power outages, intermittently loss of connectivity and all kinds of anomalies. 

### Steps to reproduce

This is a method to reproduce immediately. It happend to me without a HA restart, just not immediately. 

1. Set up a SMB storage for frigate in HA on a network share
2. take the SMB storage offline
3. Restart HA
4. After the restart HA is unable to mount /media/frigate
5. Frigate starts to write to /media/frigate on the local storage
6. any further reboots will not restore recording to SMB share since /media/frigate is already occupied
7. HA disk will run full quicly
8. HA is bricked

### Version

0.14.1-f4f3cfa

### Frigate config file

```yaml
-
```

### docker-compose file or Docker CLI command

```yaml
-
```

### Relevant Frigate log output

```shell
-
```

### Relevant go2rtc log output

```shell
-
```

### Operating system

HassOS

### Install method

HassOS Addon

### Network connection

Wired

### Camera make and model

Dahua

### Screenshots of the Frigate UI's System metrics pages

see above

**Comment by NickM-27** [comment]
This is not a bug, and it seems you are misunderstanding how this is working. 

Frigate runs in a docker container, it has no knowledge of what kind of storage it is writing to, if that storage disconnects, or anything of the sort. That is all handled by the host. 

In docker config it is trivial to configure a container to stop if the network storage disconnects. If you're running as an addon then this would be a feature request for the home assistant supervisor

**Reply by DReffects** [reply]
(I run frigate as an Addon to Home Assistant)

I am sorry to say that I cannot follow. Frigate breaks home assistant. I'd consider this a bug. As demonstrated, a cardinal duty of any surveillance system, the recording itself, is not fault proof to an extent one might expect from such a critical system. 

As an end user I have no insight in the inner workings of frigate. But I see that it breaks recording and makes home assistant unusable when a feature, outlined in the frigate documentation, (recording to remote storage https://docs.frigate.video/guides/ha_network_storage/ ) is being used.

Also, I don't see how your suggestion would solve the loss of recordings. If HA supervisor stops frigate when external storage becomes unavailable, the situation gets even worse compared to now since there's neither the ability to watch live video nor a temporary contuation of recordings to the local HA system disk until this disk gets full. 

In short: in the current state frigate is not really realiable when using remote storage. Plus, it endangeres the Home Assistant Installation. It does not even stop writing to disk when the disk is near 100%. 

If not a bug, then a most emphatically (feature) request to rethink and harden reliablity in these regards.
Frigate is awesome. But in order to replace NVR solutions like Protect from Ubiquiti it needs to be reliable.

**Reply by NickM-27** [reply]
> As an end user I have no insight in the inner workings of frigate.

You made specific statements about how you believe Frigate is working and decisions that Frigate is making, I am pointing out that your statements are incorrect and in general Frigate does not have knowledge of this occurring or have any part in the decision of where the storage is written on the host. 

HA OS reconnecting to a NAS after it is back online is already a known problem in their implementation and not something Frigate controls at all. 

> Plus, it endangeres the Home Assistant Installation. It does not even stop writing to disk when the disk is near 100%.

Frigate already has logic to ensure that the storage that is being written to is not full and runs emergency cleanup when it gets close [see the docs](https://docs.frigate.video/configuration/record#will-frigate-delete-old-recordings-if-my-storage-runs-out). If you believe this was not happening then we would need debug logs for `frigate.storage` so we can see why it was not able to recover.

> Also, I don't see how your suggestion would solve the loss of recordings. If HA supervisor stops frigate when external storage becomes unavailable, the situation gets even worse compared to now since there's neither the ability to watch live video nor a temporary contuation of recordings to the local HA system disk until this disk gets full.

It's not just about stopping frigate, many configurations are possible. It is also possible to merge the recordings that are stored on the host (in the docker volume layer) with the existing storage once it is back online. So along with improving the network storage system, HA supervisor could recover itself. All of this is already possible when Frigate is installed with docker directly. https://docs.docker.com/engine/storage/volumes/#create-cifssamba-volumes

Additionally, stopping Frigate would still ensure that the host storage is not written to and then Frigate can be restarted when the storage comes back online, which in many cases would be a small amount of time. If the network storage was down for a large amount of time then most recordings wouldn't be saved anyway as it would be overwritten given that the host storage is so small in many cases. 

In any case, Frigate does not have knowledge of the implementation details of the storage given that it runs inside of a container (addon).

**Reply by DReffects** [reply]
Thank you very much for going into such detail with this :-)
My only intention is to improve frigate - especially for enthusiastic "end users" like me. Please excuse my lack of knowledge. 

> HA OS reconnecting to a NAS after it is back online is already a known problem in their implementation and not something Frigate controls at all.

If it is a known issue with the primary environment frigate runs in and is designed for, i'd suggest that the entire storage process undergoes a sort of "harding". If frigate would be aware of storage methods and destinations it would be able to be more robust in case of failures. Or is this a limitation imposed by home assistant or the docker thing when running as an addon/container?

Am I correct in the assumtion, that if /media/frigate does not exist, **frigate** (not the host) would create this location? 
I am thinking about a fail safe with a configuration variable $external_storage == true/false check like

if $external_storage == true AND is_avail($storagepath) == false
     THEN emergency_stop()
     ELSE  create_directory()

Or, instead of stopping to record use /media/frigate_fallback instead. 

if $external_storage == true AND is_avail($storagepath) == false
     THEN record_to_fallback_location(); ALERT ("omg external storage missing")
     ELSE  create_directory()

I was able to reproduce the issue by restarting frigate while external storage was down and it started to create the frigate folder on the local disk - which then prevents the HA Host system from creating the SMB share under this location since it's already occupied.

So if you have let's say a power outage and frigate / Home assistant comes back online before the network storage (which is rather likely since a small NUC with HA boots in like 15 seconds while a larger storage system takes at least 2 minutes to even spin up the drives) you have this error. Even when recording to a simple NAS this is very likely as they tend to boot rather slowly. 

If a check had occured if external storage was being used/expected then you could simply not start recoding when the /media/frigate folder is not present, make the user aware of this in the GUI / Push alerts and use a secondary location, if this is possible. 

> Frigate already has logic to ensure that the storage that is being written to is not full and runs emergency cleanup when it gets close [see the docs](https://docs.frigate.video/configuration/record#will-frigate-delete-old-recordings-if-my-storage-runs-out). If you believe this was not happening then we would need debug logs for frigate.storage so we can see why it was not able to recover.

I guess in my case frigate was unable to recognize that the previously stored data was not available (see my screenshot with the 1888% usage reported) and therefor was also not able to delete the oldest 2 hours of footage to free up space. 

How do i enable debug logs for storage? Is it frigate.record.maintainer: debug?

If the only fail-safe routine/check within frigate is to delete old recordings I'd suggest to combine this with a hard record shut off threshold, preferably with its own configuration variable to be fitting for different environments. (if diskspace < 4GB then stop recording + alert)

Also, if frigate detects disk usage above 100% as in my case it would be a clear indicator that something went terribly wrong with storage.

**Reply by NickM-27** [reply]
> If it is a known issue with the primary environment frigate runs in and is designed for,

I don't know where you got that idea but that is incorrect, the docs specifically recommend running in native docker on a debian host. 

https://docs.frigate.video/frigate/installation#operating-system

> Am I correct in the assumtion, that if /media/frigate does not exist, frigate (not the host) would create this location?
I am thinking about a fail safe with a configuration variable $external_storage == true/false check like

you are incorrect, docker does this before Frigate even begins starting up.

**Reply by NickM-27** [reply]
> I guess in my case frigate was unable to recognize that the previously stored data was not available (see my screenshot with the 1888% usage reported) and therefor was also not able to delete the oldest 2 hours of footage to free up space.

that shouldn't interfere with the emergency cleanup, it specifically checks if the files it deleted were actually deleted before counting it as storage that was cleared. 

> If the only fail-safe routine/check within frigate is to delete old recordings I'd suggest to combine this with a hard record shut off threshold, preferably with its own configuration variable to be fitting for different environments. (if diskspace < 4GB then stop recording + alert)

Like I said and pointed to the docs, this already exists. Except it doesn't stop recording, it just deletes the oldest recordings that are available. 

> How do i enable debug logs for storage? Is it frigate.record.maintainer: debug?

The last minute storage check is `frigate.storage`, `frigate.record.maintainer` is the part that saves recordings to the disk

**Reply by DReffects** [reply]
Hi Nick,

> I don't know where you got that idea but that is incorrect, the docs specifically recommend running in native docker on a debian host.
I got that from the main page here:
![image](https://github.com/user-attachments/assets/32233699-3c60-46bc-b0c1-376ae43b441f)
My association as an end-user is, if something is "designed for Home Assistant" it will be installed from the end-user's perspective, therefor as an addon to Home Asistant. The linked doc referes to https://docs.frigate.video/integrations/home-assistant which then goes on to tell me that the  "best way to integrate with Home Assistant" is to use the official integration.
Since I have no clue about what a docker container does or is or if Debian is some sort of ex girlfriend or cookie (coming from the end-user perspective), I just followed the step by step guide from the docs and ended up with frigate as a HA addon plus the integration from HACS. If that's not the prefered way I am sorry that I missed that. From my point of view it is addon that broke home assistant ;-)

> you are incorrect, docker does this before Frigate even begins starting up.
Damn :-( 

What would you recommend to make this more stable for my use case? I do not have the hardware to set up another box just for frigate. Can I improve handling of storage from Home Assistant somehow? It was a huge pain in the behind to get HA working again after the disk got full and it's bound to happen again sooner or later since I will update ubiquiti switches or reboot my storage system from time to time. 

> that shouldn't interfere with the emergency cleanup, it specifically checks if the files it deleted were actually deleted before counting it as storage that was cleared.

odd. I will enable storage debug logs and test it when i've got spare time. Thanks!!

**Reply by NickM-27** [reply]
The integration works regardless of how Frigate is installed, and one of the main advantages is using it with home assistant. 

> What would you recommend to make this more stable for my use case? I do not have the hardware to set up another box just for frigate. Can I improve handling of storage from Home Assistant somehow? It was a huge pain in the behind to get HA working again after the disk got full and it's bound to happen again sooner or later since I will update ubiquiti switches or reboot my storage system from time to time.

I am not super familiar with what is available to HA OS with a network storage configured but I can think of a few different ideas assuming HA doesn't improve this.

1. if there is a sensor that represents the state of the network storage then you could turn off recording for cameras in Frigate when this goes offline.
2. You could also use the ubiquiti intgration to know when that goes offline and turn off frigate or just recording.

**Reply by DReffects** [reply]
Thanks, I'll look into that and get back to you when I was able to test the issue with emergency cleanup not working.

**Reply by mofonics** [reply]
Did you ever figure this out? I am experiencing the same problem. Thanks!

**Reply by JoeyGE0** [reply]
+1 - especially for the **Home Assistant add-on + external storage** case where Frigate (or the mount layer) ends up creating a **local `/media/frigate` stub** when the real disk/NAS isn't ready.

## Setup
- Home Assistant OS 18.0 (amd64)
- Frigate `0.18.0-90761a5` (HA Full Access add-on)
- 6TB USB HDD via **Samba NAS 2** → CIFS mount at `/media/frigate` (`//172.30.32.1/frigate`)
- `frigate.db` on `/config` (local SSD)

## What happened
After reboot / mount race:
1. External storage wasn't mounted yet (or Samba NAS hadn't registered the share)
2. A **local** `/media/frigate` directory appeared on the boot SSD
3. Frigate wrote ~**89GB** of `recordings/` there while I thought it was on the 6TB disk
4. When Samba NAS 2 later tried to mount, Supervisor failed with:
   `Cannot mount bind_frigate at /data/media/frigate because it is not empty`
5. Had to manually `rm -rf /media/frigate` (SSD stub only), remount Samba NAS, then restart Frigate

## Expected behaviour
If configured external/media storage is unavailable:
- **Do not** auto-create `/media/frigate/recordings|clips|exports` on local disk
- Either refuse to start recording (live view/detect can continue) **or** fail startup with a clear error
- Show a visible warning in the UI when storage is unavailable

## Why this matters on HA
Maintainers have noted Docker/HA controls mounting — agreed. But from a user perspective Frigate still ends up silently filling the boot SSD. A Frigate-side guard (e.g. detect unexpected local path / missing CIFS mount / `fail_if_unavailable` config) would help even when we can't fix Supervisor mount ordering.

## Current workaround
- Frigate add-on: **Start on boot OFF**
- Start Samba NAS 2 first, confirm `df -h /media/frigate` shows CIFS ~6TB (not nvme)
- Automation to start Frigate only after mount is healthy
